enumerable_custom.rb

```diff
@@ -0,0 +1,117 @@
+module Enumerable
+  def my_each
+    array = to_a
+    if block_given?
+      array.each { |i| yield(i) }
+    else
+      to_enum(:my_each)
+    end
+  end
+
+  def my_each_with_index
+    array = to_a
+    i = 0
+    if block_given?
+      until i == array.length
+        yield(array[i], i)
+        i += 1
+      end
+    else
+      to_enum(:my_each_with_index)
+    end
+  end
+
+  def my_select(*)
+    if block_given?
+      array = []
+      my_each { |item| array << item if yield(item) }
+    else
+      to_enum(:my_select)
+    end
+    array
+  end
+
+  def my_all?
+    if block_given?
+      condition = true
+      my_each { |item| condition = false if yield(item) == false }
+    else
+      to_enum(:my_all?)
+    end
+    condition
+  end
+
+  def my_any?
+    if block_given?
+      condition = false
+      my_each { |item| condition = true if yield(item) == true }
+    else
+      to_enum(:my_any?)
+    end
+    condition
+  end
+
+  def my_none?
+    condition = true
+    my_each { |item| condition = false if yield(item) }
+    condition
+  end
+
+  # rubocop:disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+  def my_count(*args)
+    array = is_a?(Range) ? to_a : self
+    return array.length if !block_given? && args.empty?
+
+    number = 0
+    if block_given?
+      num_elements = []
+      array.my_each do |item|
+        if yield(item)
+          num_elements << item
+          number = num_elements.length
+        end
+      end
+    else
+      array.my_each do |item|
+        unless args.empty?
+          number += 1 if item == args[0]
+        end
+      end
+    end
+    number
+  end
+
+  # rubocop:enable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+  def my_map
+    if block_given?
+      map = []
+      if proc
+        my_each { |item| map << proc.call(item) }
+      else
+        map << yield(item)
+      end
+    else
+      to_enum(:my_map)
+    end
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_map returns a new array with the results of running block once for every element in enum.
  ```ruby
    array = [1,2,5,4,7]
    block = proc { |num| num < 5 }
    p array.my_map?(&block) == array.map?(&block) # false
  ```